File: solutions/cpp/cyclic-right-shift.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Linked_list_prototype.h"

using std::cout;
using std::endl;
using std::make_shared;
using std::shared_ptr;

// @include
shared_ptr<ListNode<int>> CyclicallyRightShiftList(
    shared_ptr<ListNode<int>> L, int k) {
  if (!L || !k) {
    return L;
  }

  // Computes n, the length of L, and stores the tails of the list.
  auto tail = L;
  int n = 1;
  while (tail->next) {
    ++n, tail = tail->next;
  }

  tail->next = L;  // Makes a cycle by connecting the tail to the head.
  int steps = n - k % n;
  auto prev = tail;
  while (steps--) {
    prev = prev->next;;
  }
  L = prev->next;
  prev->next = nullptr;
  return L;
}
// @exclude

int main(int argc, char* argv[]) {
  shared_ptr<ListNode<int>> L;
  L = make_shared<ListNode<int>>(ListNode<int>{
      1, make_shared<ListNode<int>>(ListNode<int>{
             2, make_shared<ListNode<int>>(ListNode<int>{3, nullptr})})});
  auto result = CyclicallyRightShiftList(L, 2);
  assert(result->data == 2 && result->next->data == 3 && result->next->next->data == 1 && !result->next->next->next);
  while (result) {
    cout << result->data << endl;
    result = result->next;
  }
  return 0;
}

File: solutions/python/cyclic_right_shift.py
from linked_list_prototype import ListNode


# @include
def cyclically_right_shift_list(L, k):
    if not L:
        return L

    # Computes the length of L and the tail.
    tail, n = L, 1
    while tail.next:
        n += 1
        tail = tail.next

    k %= n
    if k == 0:
        return L

    tail.next = L  # Makes a cycle by connecting the tail to the head.
    steps_to_new_head, new_tail = n - k, tail
    while steps_to_new_head:
        steps_to_new_head -= 1
        new_tail = new_tail.next

    new_head = new_tail.next
    new_tail.next = None
    return new_head
# @exclude


def simple_test():
    L = ListNode(1)
    assert cyclically_right_shift_list(L, 2) is L
    L.next = ListNode(2)
    result = cyclically_right_shift_list(L, 2)
    assert result is L
    result = cyclically_right_shift_list(L, 3)
    assert result.next is L


def main():
    simple_test()
    L = ListNode(1, ListNode(2, ListNode(3, None)))
    result = cyclically_right_shift_list(L, 2)
    assert result.data == 2 and result.next.data == 3 and result.next.next.data == 1 and not result.next.next.next
    while result:
        print(result.data)
        result = result.next


if __name__ == '__main__':
    main()
